airbyte-integrations/connectors/source-snapchat-marketing/source_snapchat_marketing/source.py

```diff
@@ -0,0 +1,354 @@
+            return []
+
+        result = [{slice_key_name: cls_id} for cls_id in cls_ids]
+        auxiliary_id_map[slice_key_name] = result
+        return result
+
+
+class Organizations(SnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#organizations """
+
+    data_field = ["organizations", "organization"]
+
+    def path(self, **kwargs) -> str:
+        return "me/organizations"
+
+
+class AdAccounts(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-ad-accounts """
+
+    data_field = ["adaccounts", "adaccount"]
+    cls_object = Organizations
+    slice_key_name = "organization_id"
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        return f"organizations/{stream_slice['organization_id']}/adaccounts"
+
+
+class Creatives(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-creatives """
+
+    data_field = ["creatives", "creative"]
+    cls_object = AdAccounts
+
+
+class Media(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-media """
+
+    data_field = ["media", "media"]
+    cls_object = AdAccounts
+
+
+class Campaigns(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-campaigns """
+
+    data_field = ["campaigns", "campaign"]
+    cls_object = AdAccounts
+
+
+class Ads(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-ads-under-an-ad-account """
+
+    data_field = ["ads", "ad"]
+    cls_object = AdAccounts
+
+
+class Adsquads(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-ad-squads-under-an-ad-account """
+
+    data_field = ["adsquads", "adsquad"]
+    cls_object = AdAccounts
+
+
+class Segments(IncrementalSnapchatMarketingStream):
+    """ Docs: https://marketingapi.snapchat.com/docs/#get-all-audience-segments """
+
+    data_field = ["segments", "segment"]
+    cls_object = AdAccounts
+
+
+class SnapchatAdsOauth2Authenticator(Oauth2Authenticator):
+    """Request example for API token extraction:
+    curl -X POST https://accounts.snapchat.com/login/oauth2/access_token \
+      -d "refresh_token={refresh_token}" \
+      -d "client_id={client_id}" \
+      -d "client_secret={client_secret}"  \
+      -d "grant_type=refresh_token"  \
+    """
+
+    def __init__(self, config):
+        super().__init__(
+            token_refresh_endpoint="https://accounts.snapchat.com/login/oauth2/access_token",
+            client_id=config["client_id"],
+            client_secret=config["secret"],
+            refresh_token=config["refresh_token"],
+        )
+
+    def refresh_access_token(self) -> Tuple[str, int]:
+        response_json = None
+        try:
```

Comment:
response_json is not used after try/except. Remove this definition

Reply:
if exception will be thrown on this command in try:
`response = requests.request(method="POST", url=self.token_refresh_endpoint, data=self.get_refresh_request_body())`
the response_json will not be assigned and another exception happends inside except clause. That's why I have this `response_json = None` statement

Reply:
I see, ok